src/engraving/dom/edit.cpp

```diff
@@ -3568,138 +3590,134 @@ void Score::deleteAnnotationsFromRange(Segment* s1, Segment* s2, track_idx_t tra
-                    f = cr1->ticks();
-                } else {
-                    removeChordRest(cr1, true);
-                    f += cr1->ticks();
-                }
-            }
-            if (f.isValid() && !f.isZero()) {
-                if (fullMeasure) {
-                    // handle this as special case to be able to
-                    // fix broken measures:
-                    Staff* staff = Score::staff(track / VOICES);
-                    for (Measure* m = s1->measure(); m; m = m->nextMeasure()) {
-                        Fraction tick3   = m->tick();
-                        Fraction ff = m->stretchedLen(staff);
-                        Rest* r = setRest(tick3, track, ff, false, 0);
-                        crs.push_back(r);
-                        if (s2 && (m == s2->measure())) {
-                            break;
-                        }
-                    }
-                } else {
-                    std::vector<Rest*> rests = setRests(tick, track, f, false, currentTuplet);
-                    for (Rest* r : rests) {
-                        crs.push_back(r);
-                    }
+            if (e->isMeasureRepeat()) {
+                deleteItem(e);
+                continue;
+            }
+
+            if (currentTuplet == cr1->tuplet()) {
+                removeChordRest(cr1, true);
+                f += cr1->ticks();
+                continue;
+            }
+
+            if (f.isValid() && !fullMeasure) { // Set rests for the previous tuplet we were dealing with
+                setRest(tick, track, f, false, currentTuplet);
+            }
+
+            if (Tuplet* topTuplet = topTupletInRange(cr1->tuplet(), startTick, endTick, fullMeasure)) {
+                cmdDeleteTuplet(topTuplet, false);
+                tick = topTuplet->tick();
+                f = topTuplet->ticks();
+                currentTuplet = topTuplet->tuplet();
+                continue;
+            }
+
+            // Not deleting a complete tuplet
+            tick = cr1->tick();
+            currentTuplet = cr1->tuplet();
+            removeChordRest(cr1, true);
+            f = cr1->ticks();
+        }
+
+        if (!f.isValid() || f.isZero()) {
+            continue;
+        }
+
+        if (fullMeasure) {
+            // handle this as special case (fix broken measures):
+            const Staff* staff = Score::staff(track / VOICES);
+            for (const Measure* m = s1->measure(); m; m = m->nextMeasure()) {
+                const Fraction length = m->stretchedLen(staff);
+                Rest* r = setRest(m->tick(), track, length, false, nullptr);
+                crs.push_back(r);
+                if (s2 && m == s2->measure()) {
+                    break;
                 }
             }
+        } else {
+            const std::vector<Rest*> rests = setRests(tick, track, f, false, currentTuplet);
+            for (Rest* r : rests) {
+                crs.push_back(r);
```

Comment:
I think, something like crs.insert(crs.end(), rests.begin(), rests.end()) would also work here. Or we can even use insert with std::move to avoid creating a copy